imageio/core/imopen.py

```diff
@@ -0,0 +1,366 @@
+import numpy as np
+
+from .format import FormatManager, MODENAMES
+from .request import IOMode, Request
+
+
+class imopen:
+    """Open a URI and return a plugin instance that can read/write the URIs content.
+
+    ``imopen`` takes a URI and searches for a plugin capable of opening it. Once
+    a suitable plugin is found, a plugin instance is created that implements the
+    imageio API to interact with the image data. The search can be skipped using
+    the optional ``plugin="plugin_name"`` argument.
+
+    Notes
+    -----
+
+    For library maintainers: If you call imopen from inside the library, you
+    will first need to first create an instance, i.e. ``imopen()(uri, ...)``
+    (Notice the double parentheses).
+
+    Examples
+    --------
+
+    >>> import imageio.v3 as iio
+    >>> with iio.imopen("/path/to/image.png", "r") as file:
+    >>>     im = file.read()
+
+    >>> with iio.imopen("/path/to/output.jpg", "w") as file:
+    >>>     file.write(im)
+
+    """
+
+    _known_plugins = dict()
+    _legacy_format_manager = FormatManager()
+
+    def __call__(
+        self,
+        uri,
+        io_mode: str,
+        *,
+        plugin: str = None,
+        search_legacy_only: bool = True,
+        **kwargs,
+    ):
+        """Instantiate a plugin capable of interacting with the given URI
+
+        Parameters
+        ----------
+        uri : {str, pathlib.Path, bytes, file}
+            The resource to load the image
+            from, e.g. a filename, pathlib.Path, http address or file object,
+            see the docs for more info.
+        io_mode : {str}
+            The mode to open the file with. Possible values are::
+
+                ``r`` - open the file for reading
+                ``w`` - open the file for writing
+
+            Depreciated since v2.9:
+            A second character can be added to give the reader a hint on what
+            the user expects. This will be ignored by new plugins and will
+            only have an effect on legacy plugins. Possible values are::
+
+                ``i`` for an image,
+                ``I`` for multiple images,
+                ``v`` for a volume,
+                ``V`` for multiple volumes,
+                ``?`` for don't care (default)
+
+        plugin : {str, None}
+            The plugin to be used. If None (default), performs a search for a
+            matching plugin.
+        search_legacy_only : {bool}
+            If true (default), and ``plugin=None`` then only legacy plugins
+            (v2.9 and prior) are searched. New plugins (v3.0+) are skipped.
+        **kwargs : {any}
+            Additional keyword arguments will be passed to the plugin instance.
+        """
+
+        request = Request(uri, io_mode)
+        plugin_instance = None
+
+        if plugin is not None:
+            try:
+                plugin_instance = self._known_plugins[plugin]
```

Comment:
Yes, we probably want to re-introduce selection by file extension. I would want to have that in a dedicated PR.

Keeping in mind https://github.com/imageio/imageio/issues/636 , I would like to tweak the way it works now. I'd like to have a list of formats associated with an extension instead of just a single one. If that lookup fails, I would then like to fall back to checking all the formats that haven't been checked (in case it was missed or the extension is wrong).

Reply:
sounds good!